contracts/protocore/Protocore.sol

```diff
@@ -310,6 +317,216 @@ contract Protocore is MosaicVersion, CoconsensusModule {
+     *       targetFinalisation of the target checkpoint is set to justified.
+     * \post If the link length is equal to the epoch length (finalisation link)
+     *       then marks targetFinalisation of the link pointed by
+     *       parentVoteMessageHash of the given link as finalised once quorum
+     *       is reached.
+     * \post Calls coconsensus if the source checkpoint of the link gets
+     *       finalised.
+     */
+    function registerVoteInternal(
+        bytes32 _voteMessageHash,
+        bytes32 _r,
+        bytes32 _s,
+        uint8 _v
+    )
+        internal
+    {
+        require(
+            _voteMessageHash != bytes32(0),
+            "Vote message hash is 0."
+        );
+
+        Link storage link = links[_voteMessageHash];
+
+        require(
+            link.targetFinalisation >= CheckpointFinalisationStatus.Registered,
+            "The given link status is at least reported."
+        );
+
+        require(
+            link.proposedMetablockHeight.add(1) >= openKernelHeight,
+            "Link height inclusion principle has surpassed."
+        );
+
+        address validator = ecrecover(_voteMessageHash, _v, _r, _s);
+
+        require(
+            validator != address(0),
+            "Validator must not be null."
+        );
+
+        bool quorumReached = countVoteForForwardValidatorSets(_voteMessageHash, link, validator);
+        insertForwardValidatorVote(_voteMessageHash, validator);
+
+        if (quorumReached) {
+            justifyLink(link);
+        }
+    }
+
+
+    /* Private Functions */
+
+    /**
+     * @notice calculateQuorum() function calculates quorum of the given
+     *         count.
+     *         The super majority numerator and dominator must be the same
+     *         as in the corresponding core.
+     */
+    function calculateQuorum(uint256 _count)
+        private
+        pure
+        returns (uint256 quorum_)
+    {
+        quorum_ = _count
+            .mul(CORE_SUPER_MAJORITY_NUMERATOR)
+            .div(CORE_SUPER_MAJORITY_DENOMINATOR);
+    }
+
+    function countVoteForForwardValidatorSets(
+        bytes32 _voteMessageHash,
+        Link storage _link,
+        address _validator
+    )
+        private
+        returns (bool quorumReached_)
+    {
+        bool quorumForwardValidatorSet = false;
+        if (canVoteInForwardValidatorSet(
+            _voteMessageHash,
+            openKernelHeight,
+            openKernelHeight,
+            _validator)) {
+            _link.fvsVoteCount[openKernelHeight] = _link.fvsVoteCount[openKernelHeight].add(1);
+            quorumForwardValidatorSet = (_link.fvsVoteCount[openKernelHeight] >= fvsQuorums[openKernelHeight]);
+        }
+
+        bool quorumRearValidatorSet = true;
+        if (openKernelHeight > uint256(0)) {
+            uint256 previousHeight = openKernelHeight.sub(1);
+            if (canVoteInForwardValidatorSet(
+                _voteMessageHash,
+                openKernelHeight,
```

Comment:
openKernelHeight which is _votedHeight in canVoteInForwardValidatorSet should be openKernelHeight - 1 to catch double voting.

Reply:
corrected, moved global requirement up; and `canVoteInForwardValidatorSet` now only has one height